File: 2024/day16.py
import AOCInit
import util

import heapq as hq
from collections import defaultdict

if __name__ != '__main__':
    exit()

inp = util.getInput(d=16, y=2024)

graph = inp.splitlines()
n = len(graph)
m = len(graph[0])
start = divmod(inp.index('S'), m + 1)
end = divmod(inp.index('E'), m + 1)
dirs = (
        (0, 1),
        (-1, 0),
        (0, -1),
        (1, 0)
)

# part 1

minCost = -1
visited: dict[tuple[int, int, int], int] = dict()
h = [(0, *start, 0, *start, 0)]
prevStates = defaultdict(set)
buff = []
while len(h) != 0:
    cost, x, y, d, ox, oy, od = hq.heappop(h)
    currState = (x, y, d)
    if visited.get((x, y, d), cost) == cost:
        prevStates[(x, y, d)].add((ox, oy, od))
    if (x, y) == end:
        if minCost == -1:
            print(cost)
            minCost = cost
        buff.append((x, y, d))
        continue
    if (x, y, d) in visited:
        continue
    if minCost != -1 and cost > minCost:
        break
    visited[(x, y, d)] = cost
    # r1
    newState = (x, y, (d + 1) % 4)
    if visited.get(newState, cost + 1000) == cost + 1000:
        hq.heappush(h, (cost + 1000, *newState, *currState))
    # r2
    newState = (x, y, (d + 3) % 4)
    if visited.get(newState, cost + 1000) == cost + 1000:
        hq.heappush(h, (cost + 1000, *newState, *currState))
    # f
    xx = x + dirs[d][0]
    yy = y + dirs[d][1]
    newState = (xx, yy, d)
    if graph[xx][yy] != '#' and visited.get(newState, cost + 1) == cost + 1:
        hq.heappush(h, (cost + 1, *newState, *currState))


# part 2
tiles = set()
while len(buff) != 0:
    x, y, d = buff.pop()
    tiles.add((x, y))
    if (x, y, d) != (*start, 0):
        buff.extend(prevStates[(x, y, d)])
print(len(tiles))

File: 2024/day16.cpp
#include <queue>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

#include "../util.hpp"

using pr = std::pair<int, int>;

constexpr int DIRS[4][2] = {
    {0, 1},
    {1, 0},
    {0, -1},
    {-1, 0}
};

int main(int, char**){
    std::string inp = util::getInput(2024, 16);
    std::vector<std::string> graph = util::splitline(inp);
    int n = graph.size(), m = graph[0].size();
    int sidx = inp.find('S'), eidx = inp.find('E');
    sidx = (sidx / (m + 1)) * m + (sidx % (m + 1));
    eidx = (eidx / (m + 1)) * m + (eidx % (m + 1));

    // part 1
    int minCost = -1;
    std::priority_queue<pr, std::vector<pr>, std::greater<pr>> pq;
    pq.push({0, sidx * 4 * 4});
    std::unordered_map<int, int> visited, prevActions;
    std::vector<int> buff;
    while (!pq.empty()){
        auto [cost, x] = pq.top();
        pq.pop();
        int lastAct = x % 4;
        x /= 4;
        int currState = x, d = x % 4;
        x /= 4;
        int y = x % m;
        x /= m;
        if (util::getWithDefault(currState, visited, cost) == cost){
            prevActions[currState] |= (1 << lastAct);
        }
        if (currState / 4 == eidx){
            if (minCost == -1){
                util::output(cost);
                minCost = cost;
            }
            buff.push_back(currState);
            continue;
        }
        if (util::contains(currState, visited)){
            continue;
        }
        if (minCost != -1 && cost > minCost){
            break;
        }
        visited[currState] = cost;
        int newState = (x * m + y) * 4 + (d + 1) % 4;
        if (util::getWithDefault(newState, visited, cost + 1000) == cost + 1000){
            pq.push({cost + 1000, newState * 4 + 1});
        }
        newState = (x * m + y) * 4 + (d + 3) % 4;
        if (util::getWithDefault(newState, visited, cost + 1000) == cost + 1000){
            pq.push({cost + 1000, newState * 4 + 2});
        }
        x += DIRS[d][0];
        y += DIRS[d][1];
        newState = (x * m + y) * 4 + d;
        if (graph[x][y] != '#' && util::getWithDefault(newState, visited, cost + 1) == cost + 1){
            pq.push({cost + 1, newState * 4 + 3});
        }
    }

    // part 2
    std::vector<bool> tiles(n * m, false), added(n * m * 4, false);
    while (!buff.empty()){
        int state = buff.back();
        buff.pop_back();
        tiles[state / 4] = true;
        if (state != sidx * 4 && !added[state]){
            added[state] = true;
            int d = state % 4, coor = state / 4;
            int y = coor % m, x = coor / m;
            for (int i = 1; i <= 3; ++i){
                if ((prevActions[state] & (1 << i)) == 0){
                    continue;
                }
                switch (i){
                    case 1:
                        buff.push_back((x * m + y) * 4 + (d + 3) % 4);
                        break;
                    case 2:
                        buff.push_back((x * m + y) * 4 + (d + 1) % 4);
                        break;
                    case 3:
                        buff.push_back(((x - DIRS[d][0]) * m + y - DIRS[d][1]) * 4 + d);
                        break;
                }
            }
        }
    }
    util::output(util::sum<int>(tiles));

    return 0;
}
